### PR TITLE
feat: Creating twitter poll

### PR DESCRIPTION
**Pull request by dudu0506**
# Pull Request

## Related Issue

Closes FW-0000

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: src/providers/types/Poll.ts [hunk 1]
```diff
@@ -8,6 +8,13 @@ export interface Poll {
     options: string[];
 }
 
+export interface PollPureOption {
+    id: string;
+    text: string;
+}
+
+export type PurePoll = Pick<Poll, 'validInDays'> & { options: PollPureOption[] };
```
Comment (guanbinrui) [inline src/providers/types/Poll.ts:16]: Please apply modifications on `Poll`.

File: src/components/Poll/PollCard.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                            'relative mt-3 h-10 overflow-hidden rounded-[10px] border leading-10',
+                            hasVoted ? 'border-transparent' : 'border-lightMain',
+                        )}
+                        key={index}
+                    >
+                        <div
+                            className="absolute left-0 top-0 h-full rounded-[10px] bg-lightBg transition-all delay-100 duration-[800] ease-in-out"
+                            style={{ width: voteBarWidth }}
+                        />
+                        <div
```
Comment (guanbinrui) [inline src/components/Poll/PollCard.tsx:45]: Use `ClickableButton` stead.

File: src/components/Poll/PollCard.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import RightAnswerIcon from '@/assets/right-answer.svg';
+import { classNames } from '@/helpers/classNames.js';
+
+interface PollCardProps {}
+
+// TODO: Just ui component, need to implement the logic
+export function PollCard() {
+    const [userVote, setUserVote] = useState<string>();
+    const options = [
+        { label: 'Option A', votes: 5 },
```
Comment (guanbinrui) [inline src/components/Poll/PollCard.tsx:12]:
```suggestion
        { label: t`Option A`, votes: 5 },
```

File: src/components/Poll/PollButton.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { useComposeStateStore } from '@/store/useComposeStore.js';
+
+export const PollButton = memo(function PollButton() {
+    const { video, images, poll } = useCompositePost();
+    const { initPoll } = useComposeStateStore();
+
+    const pollDisabled = !!video || images.length > 0 || !!poll;
+
+    const handleCreatePoll = () => {
+        if (pollDisabled) return;
```
Comment (guanbinrui) [inline src/components/Poll/PollButton.tsx:17]: If pollDisabled is true, the button will not be clickable.

```suggestion
```

File: src/components/Poll/PollCard.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+interface PollCardProps {
+    post: Post;
+}
+
+export function PollCard({ post }: PollCardProps) {
+    const [userVote] = useState<string>();
+    const profile = useCurrentProfile(post.source);
+
+    const options = post.poll?.options || [];
```
Comment (guanbinrui) [inline src/components/Poll/PollCard.tsx:17]:
```suggestion
    const options = post.poll?.options || EMPTY_LIST;
```

File: src/components/Poll/PollCard.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    const voteCount = options.reduce((sum, current) => sum + (current.votes ?? 0), 0);
+
+    const createByCurrentUser = profile?.profileId === post.author?.profileId;
```
Comment (guanbinrui) [inline src/components/Poll/PollCard.tsx:20]:
```suggestion
    const createByCurrentUser = isSameProfile(profile, post.author);
```

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    const onValidInDaysChange = (days: number) => {
+        updatePoll({ ...poll, validInDays: days });
+    };
+    const onCustomValidInDaysChange = (value: string) => {
+        const isValidDays = NUMBER_BIGGER_THAN_ZERO.test(value);
+        const days = parseInt(value, 10);
+        if (!isValidDays || days <= defaultMaxDays) {
+            return setInputValue('');
+        }
+        onValidInDaysChange(days);
```
Comment (guanbinrui) [inline src/components/Poll/ValidInDaysSelector.tsx:40]:
```suggestion
        if (!isValidDays || days <= defaultMaxDays) {
            setInputValue('');
        } else {
            onValidInDaysChange(days);
        }
        

```

File: src/providers/types/Poll.ts [hunk 8]
```diff
@@ -1,13 +1,20 @@
...
+        label: string;
+        votes?: number;
+    }>;
     validInDays: number;
-    options: string[];
+    endDatetime?: string;
+    votingStatus?: string;
 }
 
+export type PollPureOption = Poll['options'][number];
```
Comment (guanbinrui) [inline src/providers/types/Poll.ts:16]:
```suggestion
export type PollOption = {
        id?: string;
        position?: number;
        label: string;
        votes?: number;
    };
```

File: src/providers/types/Poll.ts [hunk 9]
```diff
@@ -1,13 +1,20 @@
...
-    pollId: string;
-    startAt: string;
-    expiredAt: string;
+    id?: string;
+    options: Array<{
+        id?: string;
+        position?: number;
+        label: string;
+        votes?: number;
+    }>;
```
Comment (guanbinrui) [inline src/providers/types/Poll.ts:10]:
```suggestion
    options: PollOption[];
```

File: src/services/createPostTo.ts [hunk 10]
```diff
@@ -3,13 +3,18 @@ import { safeUnreachable } from '@masknet/kit';
...
 import type { MediaObject } from '@/types/index.js';
 
-type Options = Record<ComposeType, (images: MediaObject[], videos: MediaObject[]) => Promise<string>> & {
+type Options = Record<
+    ComposeType,
+    (images: MediaObject[], videos: MediaObject[], poll?: Post['poll']) => Promise<string>
+> & {
     uploadImages?: () => Promise<MediaObject[]>;
     uploadVideos?: () => Promise<MediaObject[]>;
+    createPoll?: () => Promise<Post['poll']>;
```
Comment (guanbinrui) [inline src/services/createPostTo.ts:17]:
```suggestion
    createPoll?: () => Promise<Poll>;
```

File: src/services/createPostTo.ts [hunk 11]
```diff
@@ -3,13 +3,18 @@ import { safeUnreachable } from '@masknet/kit';
...
 import { type SocialSource } from '@/constants/enum.js';
+import type { Post } from '@/providers/types/SocialMedia.js';
 import { type CompositePost, useComposeStateStore } from '@/store/useComposeStore.js';
 import type { ComposeType } from '@/types/compose.js';
 import type { MediaObject } from '@/types/index.js';
 
-type Options = Record<ComposeType, (images: MediaObject[], videos: MediaObject[]) => Promise<string>> & {
+type Options = Record<
+    ComposeType,
+    (images: MediaObject[], videos: MediaObject[], poll?: Post['poll']) => Promise<string>
```
Comment (guanbinrui) [inline src/services/createPostTo.ts:13]:
```suggestion
    (images: MediaObject[], videos: MediaObject[], poll?: Poll) => Promise<string>
```

File: src/services/postToTwitter.ts [hunk 12]
```diff
@@ -27,7 +27,7 @@ export async function postToTwitter(type: ComposeType, compositePost: CompositeP
     const { currentProfile } = useTwitterStateStore.getState();
     if (!currentProfile?.profileId) throw new Error(t`Login required to post on ${sourceName}.`);
 
-    const composeDraft = (postType: PostType, images: MediaObject[]) => {
+    const composeDraft = (postType: PostType, images: MediaObject[], poll?: Post['poll']) => {
```
Comment (guanbinrui) [inline src/services/postToTwitter.ts:30]:
```suggestion
    const composeDraft = (postType: PostType, images: MediaObject[], poll?: Poll) => {
```

File: src/services/postToTwitter.ts [hunk 13]
```diff
@@ -55,10 +56,17 @@ export async function postToTwitter(type: ComposeType, compositePost: CompositeP
                 file: x.file,
             }));
         },
-        compose: (images) => TwitterSocialMediaProvider.publishPost(composeDraft('Post', images)),
-        reply: (images) => {
+        createPoll: async () => {
+            if (!poll) return undefined;
```
Comment (guanbinrui) [inline src/services/postToTwitter.ts:60]:
```suggestion
            if (!poll) return;
```

File: src/helpers/createPoll.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        return POLL_OPTIONS_MAX_COUNT[Source.Lens];
+    }
+    return POLL_OPTIONS_MAX_COUNT[Source.Farcaster];
+};
+
+export const shouldShowCustomDaysInput = (availableSources: SocialSource[]) => {
+    if (availableSources.length === 1 && availableSources[0] === Source.Twitter) {
+        return false;
+    }
+    return true;
```
Comment (guanbinrui) [inline src/helpers/createPoll.ts:35]:
```suggestion
    return !(availableSources.length === 1 && availableSources[0] === Source.Twitter)
```

File: src/store/useComposeStore.ts [hunk 15]
```diff
@@ -474,6 +481,28 @@ const useComposeStateBase = create<ComposeState, [['zustand/immer', unknown]]>(
...
+                    cursor,
+                ),
+            ),
+        initPoll: (cursor) =>
+            set((state) =>
+                next(
+                    state,
+                    (post) => ({
+                        ...post,
+                        poll: createInitPurePoll(),
```
Comment (guanbinrui) [inline src/store/useComposeStore.ts:501]:
```suggestion
                        poll: createInitPoll(),
```

File: src/app/api/twitter/compose/route.ts [hunk 16]
```diff
@@ -14,6 +16,16 @@ const TweetSchema = z.object({
...
     inReplyToTweetId: z.string().optional(),
     mediaIds: z.array(z.string()).optional(),
+    poll: z
+        .object({
+            options: z.array(
+                z.object({
+                    label: z.string(),
+                }),
+            ),
+            validInDays: z.number(),
```
Comment (guanbinrui) [inline src/app/api/twitter/compose/route.ts:26]:
```suggestion
            validInDays: z.number().int().positive(),
```

File: src/app/api/twitter/compose/route.ts [hunk 17]
```diff
@@ -14,6 +16,16 @@ const TweetSchema = z.object({
     quoteTwitterId: z.string().optional(),
     inReplyToTweetId: z.string().optional(),
     mediaIds: z.array(z.string()).optional(),
+    poll: z
+        .object({
+            options: z.array(
+                z.object({
+                    label: z.string(),
```
Comment (guanbinrui) [inline src/app/api/twitter/compose/route.ts:23]: Set a maximum character count.

File: src/components/Poll/PollCreatorCard.tsx [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    const optionsMaxLength = getPollOptionsMaxLength(availableSources);
+    const addDisabled = readonly || poll.options.length >= optionsMaxLength;
+
+    const removeOption = (option: PollPureOption) => {
+        if (readonly) return;
+        const newOptions = poll.options.filter((o) => o.id !== option.id);
+        updatePoll({ ...poll, options: newOptions });
+    };
+    const addOption = () => {
+        if (addDisabled) return;
```
Comment (guanbinrui) [inline src/components/Poll/PollCreatorCard.tsx:45]:
```suggestion
```

File: src/providers/types/Poll.ts [hunk 19]
```diff
@@ -1,11 +1,18 @@
 import type { Pageable } from '@masknet/shared-base';
 
+export interface PollPureOption {
```
Comment (guanbinrui) [inline src/providers/types/Poll.ts:3]:
```suggestion
export interface PollOption {
```

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+                                        <Trans>Custom</Trans>
+                                    </span>
+                                    <form
+                                        className="flex-1 rounded-lg bg-lightBg"
+                                        onSubmit={(ev) => {
+                                            ev.preventDefault();
+                                            onCustomValidInDaysChange(inputValue);
+                                        }}
+                                    >
+                                        <input
```
Comment (guanbinrui) [inline src/components/Poll/ValidInDaysSelector.tsx:93]: max length

File: src/components/Poll/PollCreatorCard.tsx [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+                                height={20}
+                                className={readonly ? '' : 'cursor-pointer'}
+                                onClick={() => removeOption(option)}
+                            />
+                        )}
+                    </div>
+                ))}
+            </div>
+            <div className="mt-4 flex items-center justify-between">
+                <div
```
[the remark this answers is not in the document]
Comment (guanbinrui) [reply src/components/Poll/PollCreatorCard.tsx:96]: If no input value provided, we should disable the add button.

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+                                            min={defaultMaxDays + 1}
```
Comment (guanbinrui) [inline src/components/Poll/ValidInDaysSelector.tsx:94]: type number

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+                    >
+                        <Popover.Panel className="absolute bottom-full right-0 flex w-[185px] -translate-y-3 flex-col gap-2 rounded-lg bg-lightBottom py-3 text-[15px] shadow-popover dark:border dark:border-line dark:bg-darkBottom dark:shadow-none">
+                            {POLL_VALID_IN_DAYS_DEFAULT_LIST.map((day) => (
+                                <div
+                                    key={day}
+                                    className={classNames(
+                                        'h-5 cursor-pointer text-center text-base font-bold leading-5 text-lightMain md:h-7 md:leading-7',
+                                        validInDays === day ? 'bg-lightBg' : '',
+                                    )}
+                                    onClick={() => {
```
Comment (guanbinrui) [inline src/components/Poll/ValidInDaysSelector.tsx:72]: Use `<ClickableButton />` stead.

File: src/components/Compose/Editor.tsx [hunk 24]
```diff
@@ -33,6 +33,7 @@ export const Editor = memo(function Editor({ post, replying }: EditorProps) {
 
     const { chars } = post;
     const index = posts.findIndex((x) => x.id === post.id);
+    const hasPoll = !!post.poll;
```
Comment (guanbinrui) [inline src/components/Compose/Editor.tsx:36]: For a variable used only once, it's recommended to use the expression directly instead.

File: src/components/Poll/PollCard.tsx [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+export function PollCard({ post }: PollCardProps) {
+    const [userVote] = useState<string>();
+    const profile = useCurrentProfile(post.source);
+
+    const options = post.poll?.options || EMPTY_LIST;
+    const voteCount = options.reduce((sum, current) => sum + (current.votes ?? 0), 0);
+
+    const createByCurrentUser = isSameProfile(profile, post.author);
+    const isVoteStopped = post.poll?.votingStatus === 'closed';
+    const voteDisabled = isVoteStopped || !!profile || createByCurrentUser;
```
Comment (guanbinrui) [inline src/components/Poll/PollCard.tsx:24]:
```suggestion
    const voteDisabled = post.poll?.votingStatus === 'closed' || isSameProfile(profile, post.author);
```

File: src/components/Poll/PollCreatorCard.tsx [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+    const { poll } = post;
+
+    if (!poll) return null;
+
+    const optionsMaxLength = getPollOptionsMaxLength(availableSources);
+    const addDisabled = readonly || poll.options.length >= optionsMaxLength;
+
+    const removeOption = (option: PollPureOption) => {
+        if (readonly) return;
```
Comment (guanbinrui) [inline src/components/Poll/PollCreatorCard.tsx:40]: Disable the remove button if readonly.

Comment (guanbinrui) [inline src/components/Poll/PollCreatorCard.tsx:40]: Disable the remove button if readonly.

```suggestion
```

File: src/components/Poll/PollCreatorCard.tsx [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+        updatePoll({ ...poll, options: newOptions });
+    };
+    const addOption = () => {
+        if (addDisabled) return;
+        updatePoll({ ...poll, options: [...poll.options, createPollInitOption()] });
+    };
+    const onOptionChange = (option: PollPureOption, label: string) => {
+        const newOptions = poll.options.map((o) => (o.id === option.id ? { ...o, label } : o));
+        updatePoll({ ...poll, options: newOptions });
+    };
```
Comment (nuanyang233) [inline src/components/Poll/PollCreatorCard.tsx:51]: Please use `useCallback`

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+    const [inputValue, setInputValue] = useState<string>(
+        POLL_VALID_IN_DAYS_DEFAULT_LIST.includes(validInDays) ? '' : `${validInDays}`,
+    );
+    const { updatePoll } = useComposeStateStore();
+    const { availableSources } = useCompositePost();
+
+    if (!poll) return null;
+
+    const showCustomDaysInput = shouldShowCustomDaysInput(availableSources);
+    const defaultMaxDays = POLL_VALID_IN_DAYS_DEFAULT_LIST[POLL_VALID_IN_DAYS_DEFAULT_LIST.length - 1];
```
Comment (nuanyang233) [inline src/components/Poll/ValidInDaysSelector.tsx:29]: Maybe you should redefine a constant.

File: src/components/Poll/PollCard.tsx (hunk 25 above)
Comment (dudu0506) [reply src/components/Poll/PollCard.tsx]: We should also care login status

File: src/constants/poll.ts [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+
+export const POLL_OPTIONS_MAX_COUNT = {
+    [Source.Lens]: 10,
+    [Source.Farcaster]: 4,
+    [Source.Twitter]: 4,
+};
+
+export const POLL_MAX_VALID_IN_DAYS = {
+    [Source.Lens]: Infinity,
+    [Source.Farcaster]: Infinity,
```
Comment (guanbinrui) [inline src/constants/poll.ts:22]:
```suggestion
    [Source.Lens]: Number.MAX_SAFE_INTEGER,
    [Source.Farcaster]: Number.MAX_SAFE_INTEGER,

```

File: src/helpers/createPoll.ts [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+        validInDays: POLL_DEFAULT_VALID_IN_DAYS,
+        options: Array.from({ length: POLL_OPTIONS_MIN_COUNT }).map(() => createPollInitOption()),
+    };
+};
+
+export const getPollOptionsMaxLength = (availableSources: SocialSource[]) => {
+    if (availableSources.length === 1 && availableSources[0] === Source.Lens) {
+        return POLL_OPTIONS_MAX_COUNT[Source.Lens];
+    }
+    return POLL_OPTIONS_MAX_COUNT[Source.Farcaster];
```
Comment (guanbinrui) [inline src/helpers/createPoll.ts:28]: What about X?

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 31]
```diff
@@ -0,0 +1,109 @@
...
+    const validInDays = poll?.validInDays ?? 0;
+    const [inputValue, setInputValue] = useState<string>(
+        POLL_VALID_IN_DAYS_DEFAULT_LIST.includes(validInDays) ? '' : `${validInDays}`,
+    );
+    const { updatePoll } = useComposeStateStore();
+    const { availableSources } = useCompositePost();
+
+    if (!poll) return null;
+
+    const showCustomDaysInput = shouldShowCustomDaysInput(availableSources);
```
[the remark this answers is not in the document]
Comment (guanbinrui) [reply src/components/Poll/ValidInDaysSelector.tsx:28]: Not necessary here.

File: src/helpers/createPoll.ts (hunk 30 above)
Comment (guanbinrui) [inline src/helpers/createPoll.ts:28]: Please iterate `SORTED_POLL_SOURCES`.

File: src/helpers/createPoll.ts [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+export const shouldShowCustomDaysInput = (availableSources: SocialSource[]) => {
+    if (availableSources.length === 1 && availableSources[0] === Source.Twitter) {
```
Comment (guanbinrui) [inline src/helpers/createPoll.ts:32]: Please iterate `SORTED_POLL_SOURCES`.

File: src/components/Poll/ValidInDaysSelector.tsx [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+
+    return (
+        <Popover as="div" className="relative">
+            {({ open, close }) => (
+                <>
+                    <Popover.Button
+                        disabled={readonly}
+                        className="flex h-6 items-center gap-2 rounded-full border border-lightMain px-3 text-[13px] font-bold text-lightMain"
+                    >
+                        <span>{validInDays === 1 ? t`${validInDays} day` : t`${validInDays} days`}</span>
```
Comment (guanbinrui) [inline src/components/Poll/ValidInDaysSelector.tsx:53]: Use plural.

**Pull request merged**